### PR TITLE
[#2710] Improvement(gvfs): Support oauth2 client authentication in gvfs

### PR DESCRIPTION
**Pull request by xloya**
### What changes were proposed in this pull request?

Support connecting Gravitino Server through Oauth2 in gvfs. Now only supports simple auth.

### Why are the changes needed?

Fix: #2710 

### How was this patch tested?

Add the unit tests.

### CODE REVIEW

**Comment by jerryshao** [comment]
@qqqttt123 would you please help to review this?

**Comment by qqqttt123** [comment]
I will.

**Comment by qqqttt123** [comment]
Could you add some assertion for the test case?

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop3/GravitinoVirtualFileSystemConfiguration.java [hunk 1]
```diff
@@ -15,6 +15,28 @@ class GravitinoVirtualFileSystemConfiguration {
...
+  /** The configuration key for the Gravitino client auth type. */
+  public static final String FS_GRAVITINO_CLIENT_AUTH_TYPE_KEY = "fs.gravitino.client.authType";
+
+  public static final String SIMPLE_AUTH_TYPE = "simple";
+  public static final String OAUTH2_AUTH_TYPE = "oauth2";
+
+  // oauth2
+  /** The configuration key for the URI of the default OAuth server. */
+  public static final String FS_GRAVITINO_CLIENT_OAUTH2_SERVER_URI_KEY =
+      "fs.gravitino.client.oauth2.serverUri";
```
Comment (qqqttt123) [inline clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop3/GravitinoVirtualFileSystemConfiguration.java:27]: Could you add a blank line here?

Comment (xloya) [reply clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop3/GravitinoVirtualFileSystemConfiguration.java:27]: Done

File: clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    assertNotNull(response);
+    assertEquals(Oauth2MockServerBase.accessToken(), response.getAccessToken());
+    assertEquals("2", response.getIssuedTokenType());
+    assertEquals("bearer", response.getTokenType());
+    assertEquals("test", response.getScope());
+    assertEquals(1, response.getExpiresIn());
+  }
+
+  @Test
+  public void testConfiguration() throws IOException {
```
Comment (qqqttt123) [inline clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:80]: We don't need these assertions. I mean that we can add some assertion in the MockServer to verify more information.

Comment (xloya) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:80]: Do you mean to add some request and response tests, which like in the `TestOAuth2TokenProvider`?

Comment (qqqttt123) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:80]: We need to verify the requests and responses. But the new test case may be different from `TestOAuth2TokenProvider`.

File: clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    assertNotNull(response);
+    assertEquals(Oauth2MockServerBase.accessToken(), response.getAccessToken());
+    assertEquals("2", response.getIssuedTokenType());
+    assertEquals("bearer", response.getTokenType());
+    assertEquals("test", response.getScope());
+    assertEquals(1, response.getExpiresIn());
+  }
+
+  @Test
+  public void testFileSystemAuthConfigs() throws IOException {
```
Comment (xloya) [inline clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:92]: Add uts for configs check and auth exceptions.

File: clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        IllegalArgumentException.class, () -> managedFilesetPath.getFileSystem(configuration));
+
+    // set oauth scope, all configs are set
+    configuration.set(
+        GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_CLIENT_OAUTH2_SCOPE_KEY, scope);
+    assertNotNull(managedFilesetPath.getFileSystem(configuration));
+  }
+
+  @Test
+  public void testFileSystemAuthUnauthorized() {
```
Comment (qqqttt123) [inline clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:141]: Could we add a wrong token case?

File: clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java [hunk 5]
```diff
@@ -158,6 +161,33 @@ public void testFileSystemAuthUnauthorized() {
     // should throw UnauthorizedException
     assertThrows(
         UnauthorizedException.class, () -> managedFilesetPath.getFileSystem(configuration));
+
+    // test wrong client secret
```
Comment (qqqttt123) [inline clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:165]: Expired token or wrong token. I don't want to verify the OAuth server. I would like to test the Gravitino server.

Comment (xloya) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:165]: The Mock Gravitino Server in the current unit test is a relatively simple REST service without a `ServerAuthenticator` environment, I may be more inclined to put this test into the integration test.

Comment (qqqttt123) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestOauth2Client.java:165]: OK.

File: clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestSimpleClient.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.jupiter.api.BeforeAll;
+
+public class TestSimpleClient extends TestGvfsBase {
+  @BeforeAll
+  public static void setup() {
+    TestGvfsBase.setup();
+    conf.set(
+        GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_CLIENT_AUTH_TYPE_KEY,
+        GravitinoVirtualFileSystemConfiguration.SIMPLE_AUTH_TYPE);
+  }
```
Comment (qqqttt123) [inline clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestSimpleClient.java:16]: Could you verify the token from the client satisfies the expect value?

Comment (xloya) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestSimpleClient.java:16]: Do you mean check the simple auth's token?

Comment (qqqttt123) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestSimpleClient.java:16]: > Do you mean check the simple auth's token?

Yes.

Comment (xloya) [reply clients/filesystem-hadoop3/src/test/java/com/datastrato/gravitino/filesystem/hadoop3/TestSimpleClient.java:39]: Done

**Review by qqqttt123 — DISMISSED** [review]
LGTM, thanks @xloya

**Comment by xloya** [comment]
@jerryshao Please help take a look of this. Maybe we can add the security content for gvfs after this be merged.

**Comment by jerryshao** [comment]
Do you need to support kerberos?

**Comment by xloya** [comment]
> Do you need to support kerberos?

Yeah, i think add the code is simple, but how to test it in the unit tests may take some time. Recently I have been working on internal requirements and kerberos support may not be completed before 4.12.

**Comment by jerryshao** [comment]
> > Do you need to support kerberos?
> 
> Yeah, i think add the code is simple, but how to test it in the unit tests may take some time. Recently I have been working on internal requirements and kerberos support may not be completed before 4.12.

I see.

**Review by qqqttt123 — APPROVED** [review]

**Comment by qqqttt123** [comment]
Thanks @jerryshao @xloya , merge to master.

**Comment by jerryshao** [comment]
@xloya would you please add the doc in gvfs about how to use simple/oauth for gvfs?

**Comment by xloya** [comment]
> @xloya would you please add the doc in gvfs about how to use simple/oauth for gvfs?

Sure, will open a PR today.

**Pull request merged**